app/uk/gov/hmrc/helptosave/models/account/Transactions.scala

```diff
@@ -0,0 +1,151 @@
+  implicit val writes: Writes[Operation] = new Writes[Operation] {
+    override def writes(o: Operation): JsValue = JsString(o match {
+      case Credit ⇒ "credit"
+      case Debit  ⇒ "debit"
+    })
+  }
+}
+
+case class ValidNsiTransaction(
+    sequence:             Int,
+    operation:            Operation,
+    amount:               BigDecimal,
+    transactionDate:      LocalDate,
+    accountingDate:       LocalDate,
+    description:          String,
+    transactionReference: String
+)
+
+object ValidNsiTransaction {
+  private type ValidOrErrorString[A] = ValidatedNel[String, A]
+
+  def apply(nsiTransaction: NsiTransaction): ValidatedNel[String, ValidNsiTransaction] = {
+      def operationValidation(nsiOperation: String): ValidOrErrorString[Operation] =
+        if (nsiOperation === "C") { Valid(Credit) }
+        else if (nsiOperation === "D") { Valid(Debit) }
+        else { Invalid(NonEmptyList.one(s"""Unknown value for operation: "$nsiOperation"""")) }
+
+      def sequenceValidation(nsiSequence: String): ValidOrErrorString[Int] =
+        try Valid(nsiSequence.toInt)
+        catch {
+          case _: NumberFormatException ⇒ Invalid(NonEmptyList.one(s"""Can't parse sequence value as an integer: "$nsiSequence""""))
+        }
+
+    (operationValidation(nsiTransaction.operation), sequenceValidation(nsiTransaction.sequence)).mapN {
+      case (operation, sequence) ⇒
+        ValidNsiTransaction(
+          sequence,
+          operation,
+          nsiTransaction.amount,
+          nsiTransaction.transactionDate,
+          nsiTransaction.accountingDate,
+          nsiTransaction.description,
+          nsiTransaction.transactionReference
+        )
+    }
+  }
+}
+
+case class Transaction(
+    operation:            Operation,
+    amount:               BigDecimal,
+    transactionDate:      LocalDate,
+    accountingDate:       LocalDate,
+    description:          String,
+    transactionReference: String,
+    balanceAfter:         BigDecimal
+)
+
+object Transaction {
+  implicit val writes: Writes[Transaction] = Json.writes[Transaction]
+
+  def apply(vnt: ValidNsiTransaction, balanceAfter: BigDecimal): Transaction = Transaction(
+    vnt.operation,
+    vnt.amount,
+    vnt.transactionDate,
+    vnt.accountingDate,
+    vnt.description,
+    vnt.transactionReference,
+    balanceAfter
+  )
+}
+
+case class Transactions(transactions: Seq[Transaction])
+
+object Transactions {
+  implicit val writes: Writes[Transactions] = Json.writes[Transactions]
+
+  private type ValidOrErrorString[A] = ValidatedNel[String, A]
+
+  def apply(nsiTransactions: NsiTransactions): ValidatedNel[String, Transactions] = {
+    nsiTransactions.transactions.toList
+      .traverse[ValidOrErrorString, ValidNsiTransaction](ValidNsiTransaction.apply)
+      .map(sortLikeNsiWeb)
+      .map(runningBalance)
+      .map(Transactions.apply)
+  }
+
+  private implicit val eqLocalDate: Eq[LocalDate] = Eq.fromUniversalEquals
```

Comment:
you could provide a more meaningful `Eq` instance here using something like `Eq.instance[LocalDate](_.isEqual(_))`

Reply:
done